internal/controllers/ingress_controller.go

```diff
@@ -272,3 +315,44 @@ func (irec *IngressReconciler) ingressToDomains(ctx context.Context, ingress *ne
 
 	return reservedDomains, nil
 }
+
+func ingressToTunnels(ingress *netv1.Ingress) []ingressv1alpha1.Tunnel {
+	tunnels := make([]ingressv1alpha1.Tunnel, 0)
+
+	if ingress == nil || len(ingress.Spec.Rules) == 0 {
+		return tunnels
+	}
+
+	// Tunnels should be unique on a service and port basis so if they are ferenced more than once, we
+	// only create one tunnel per service and port.
+	tunnelMap := make(map[string]ingressv1alpha1.Tunnel)
+	for _, rule := range ingress.Spec.Rules {
+		if rule.Host == "" {
+			continue
+		}
+
+		for _, path := range rule.HTTP.Paths {
+			serviceName := path.Backend.Service.Name
+			servicePort := path.Backend.Service.Port.Number
+			tunnelAddr := fmt.Sprintf("%s.%s.%s:%d", serviceName, ingress.Namespace, clusterDomain, servicePort)
+			tunnelName := fmt.Sprintf("%s-%d", serviceName, servicePort)
+
+			tunnelMap[tunnelName] = ingressv1alpha1.Tunnel{
```

Comment:
making sure i'm understanding this correctly: we could have key conflicts here within one ingress, but because the value is composed only of values that are a part of the key or are consistent here, those conflicts would always have the same spec value (forwards to and labels) and so its fine?

Reply:
This might have been an early optimization, but assuming you you have an ingress where:

* `/` -> forwards to service-1 on port 80
* `/path1` -> forwards to service-2 on port 80
* `/path2` -> forwards to service-2 on port 80

The map will have:

* service-1-80 -> Tunnel for service-1 on port 80
* service-2-80 -> Tunnel for service-2 on port 80

The thinking here being we don't need to duplicate the tunnel for service-2 on port 80. Since the tunnel backends are based on selectors, we can just reference the same tunnel backend.

Let me know if that makes sense.

Reply:
yep, that makes sense. I was just clarifying that the loop at some point will have a map with `service-1-80` as a key and then override the value with a new value. But that value will be the same so it doesn't matter if the key is already present